reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerPostDecoder.java

```diff
@@ -0,0 +1,161 @@
+import io.netty.handler.codec.http.multipart.HttpPostRequestDecoder;
+import io.netty.handler.codec.http.multipart.InterfaceHttpData;
+import io.netty.handler.codec.http.multipart.InterfaceHttpPostRequestDecoder;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This decoder is able to decode a post body.
+ * <p>
+ * If you disable {@link HttpServerPostDecoder#automaticCleanup}, you <strong>HAVE TO</strong> call {@link #dispose()} after completion to release all resources.
+ */
+public class HttpServerPostDecoder implements Disposable {
+
+	private final InterfaceHttpPostRequestDecoder decoder;
+	private final HttpServerRequest request;
+
+	private boolean automaticCleanup;
+	private boolean disposed;
+
+	HttpServerPostDecoder(final HttpRequest nettyRequest, HttpServerRequest request) {
+		this.decoder = new HttpPostRequestDecoder(nettyRequest);
+		this.request = request;
+
+		this.automaticCleanup = false;
+		this.disposed = false;
+	}
+
+	@Override
+	public boolean isDisposed() {
+		return this.disposed;
+	}
+
+	/**
+	 * Returns if {@link HttpServerPostDecoder#dispose()} should be called automatically.
+	 *
+	 * @return if {@link HttpServerPostDecoder#dispose()} should be called automatically.
+	 */
+	public boolean isAutomaticCleanup() {
+		return this.automaticCleanup;
+	}
+
+	/**
+	 * Defines if {@link HttpServerPostDecoder#dispose()} should be called automatically.
+	 */
+	public HttpServerPostDecoder automaticCleanup(final boolean automaticCleanup) {
+		this.automaticCleanup = automaticCleanup;
+		return this;
+	}
+
+	/**
+	 * True if this request is a Multipart request
+	 *
+	 * @return True if this request is a Multipart request
+	 */
+	public boolean isMultipart() {
+		return this.decoder.isMultipart();
+	}
+
+	/**
+	 * Set the amount of bytes after which read bytes in the buffer should be discarded.
+	 * Setting this lower gives lower memory usage but with the overhead of more memory copies.
+	 * Use {@code 0} to disable it.
+	 */
+	public HttpServerPostDecoder discardThreshold(final int discardThreshold) {
+		this.decoder.setDiscardThreshold(discardThreshold);
+		return this;
+	}
+
+	/**
+	 * Return the threshold in bytes after which read data in the buffer should be discarded.
+	 *
+	 * @return the threshold in bytes after which read data in the buffer should be discarded.
+	 */
+	public int discardThreshold() {
+		return this.decoder.getDiscardThreshold();
+	}
+
+	/**
+	 * Return a {@link Flux} containing all {@link InterfaceHttpData}'s from the body.
+	 *
+	 * @return a {@link Flux} containing all {@link InterfaceHttpData}'s from the body.
+	 */
+	public Flux<InterfaceHttpData> getBodyHttpData() {
+		return this.request.receiveContent()
+				.flatMapIterable(content -> {
+					this.decoder.offer(content);
+
+					final List<InterfaceHttpData> data = new ArrayList<>();
+					while (this.decoder.hasNext()) data.add(this.decoder.next());
```

Comment:
Is it possible to stream the data instead of first collecting it and then stream it?

Reply:
Netty provide the `InterfaceHttpData` as soon it has enough Data (`HttpContent`), that means that Netty returns in Line 110 always false as long as it hasn't enough `HttpContent` Objects. If Netty has enough HttpContent Objects for one `InterfaceHttpData` it will provide it in the `decoder.next()` method and I collect all `InterfaceHttpData` Objects witch are decodeable with the current provided `HttpContent`. -> I am not able to stream the data directly.

[hasNext](https://netty.io/4.1/api/io/netty/handler/codec/http/multipart/HttpPostRequestDecoder.html#hasNext--)
[next](https://netty.io/4.1/api/io/netty/handler/codec/http/multipart/HttpPostRequestDecoder.html#next--)
[FileUpload](https://netty.io/4.1/api/io/netty/handler/codec/http/multipart/FileUpload.html)